File: golang/src/detectors/insecure-cryptography/InsecureCryptography.go
package main

import (
	"crypto/md5"
	"crypto/sha256"
	"fmt"
)

type User struct {
	password string
}

func (user *User) setPassword(password string) {
	user.password = password
}

// {fact rule=insecure-cryptography@v1.0 defects=1}
func insecureCryptographyNoncompliant(user *User, passText string) {
	data := []byte(passText)
	// Noncompliant: md5 used as password.
	hash := md5.Sum(data)
	password := fmt.Sprintf("%x", hash)
	user.setPassword(password)
}

// {/fact}

// {fact rule=insecure-cryptography@v1.0 defects=0}
func insecureCryptographyCompliant(user *User, passText string) {
	data := []byte(passText)
	// Compliant: sha256 used as password.
	hash := sha256.Sum256(data)
	password := fmt.Sprintf("%x", hash)
	user.setPassword(password)
}

// {/fact}

File: typescript/src/detector/critical/insecure-cryptography/insecure-cryptography.ts
// {fact rule=insecure-cryptography@v1.0 defects=1}
import https from 'https'
function insecureCryptographyNoncompliant() {
  var ciphers = [
    `TLS_DH_anon_WITH_AES_256_GCM_SHA384`,
    `TLS_AES_128_GCM_SHA256`,
    `ECDHE-ECDSA-AES128-GCM-SHA256`,
  ].join(":");
  var options = {
    hostname: "www.example.com",
    port: 443,
    path: "/",
    method: "GET",
    secureProtocol: "TLSv1_2_method",
    // Noncompliant: insecure TLS cipher suite is used.
    ciphers: ciphers,
  };

  var req = https.request(
    options,
    (res: { on: (arg0: string, arg1: (d: any) => void) => void }) => {
      res.on("data", (d: any) => {
        process.stdout.write(d);
      });
    },
  );
}
// {/fact}

// {fact rule=insecure-cryptography@v1.0 defects=0}
import https from 'https'
function insecureCryptographyCompliant() {
  var ciphers = [
    `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256`,
    `TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384`,
    `TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256`,
    `TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384`,
    `TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256`,
    `TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256`,
    `TLS_AES_128_GCM_SHA256`,
    `TLS_AES_256_GCM_SHA384`,
    "!aNULL",
    "!eNULL",
    "!NULL",
    "!DES",
    "!RC4",
    "!MD5",
  ].join(":");
  var options = {
    hostname: "www.example.com",
    port: 443,
    path: "/",
    method: "GET",
    secureProtocol: "TLSv1_2_method",
    // Compliant: secure TLS cipher suite is used.
    ciphers: ciphers,
  };

  var req = https.request(
    options,
    (res: { on: (arg0: string, arg1: (d: any) => void) => void }) => {
      res.on("data", (d: any) => {
        process.stdout.write(d);
      });
    },
  );
}
// {/fact}
